### ISSUE TITLE
Add a manual (or manual section): Customizing Pandoc

### ISSUE THREAD

**Issue by jgm** [jgm/pandoc#3288]
Topics:

- templates
- reference.docx/odt
- filters (including documentation of the JSON serialization format and [AST definition](https://github.com/jgm/pandoc-types/blob/master/Text/Pandoc/Definition.hs#L94))
- custom lua writers
- custom syntax highlighting (syntaxes, styles)

**Comment by crmackay** [comment]
- custom (lua) writers?

**Comment by jgm** [comment]
Thanks for the suggestion, I've added it to the list.

**Comment by ickc** [comment]
Suggestion: subtopics for filters: AST. Also see #3262.

Example:

Recently I knew one of the undocumented feature of the AST regarding RawInline/RawBlock. I try to make some sense out of it by doing the following tests

```bash
# Raw LaTeX
$ printf "%s\n\n" '\LaTeX' | pandoc -f markdown -t native
[Para [RawInline (Format "tex") "\\LaTeX"]]
$ printf "%s" '[Para [RawInline (Format "tex") "\\LaTeX"]]' | pandoc -t markdown -f native
\LaTeX
$ printf "%s" '[Para [RawInline (Format "latex") "\\LaTeX"]]' | pandoc -t markdown -f native
\LaTeX
$ printf "%s" '[Para [RawInline (Format "beamer") "\\LaTeX"]]' | pandoc -t markdown -f native  

$ printf "%s" '[Para [RawInline (Format "beamer") "\\LaTeX"]]' | pandoc -t beamer -f native
\begin{frame}

\end{frame}
# Raw HTML
$ printf "%s" '[RawBlock (Format "html") "<html>"]' | pandoc -t markdown -f native
<html>

$ printf "%s" '[RawBlock (Format "html5") "<html>"]' | pandoc -t markdown -f native

$ printf "%s" '[RawBlock (Format "html5") "<html>"]' | pandoc -t html5 -f native

$ printf "%s" '[RawBlock (Format "markdown") "\\LaTeX"]' | pandoc -t markdown -f native
\LaTeX
$ printf "%s" '[RawBlock (Format "markdown") "\\LaTeX"]' | pandoc -t latex -f native

```

What we see here:

- while `pandoc -t tex` is invalid, `(Format "tex")` is valid in AST and in fact is the default when raw LaTeX is used in markdown. (and "latex" is valid in both cases)

- Format valid in pandoc cli can be invalid in AST's `Format`, e.g. beamer, html5, albeit not causing any error but disappeared.

**Comment by tajmone** [comment]
Definitely, documentation on pandoc's AST is something I've been longing for a long time.

Not for using filters, but to be able to create an application that generates a valid pandoc document in JSON AST format.

For example, I've created a feature request for an Asciidoctor to pandoc AST backend:

- asciidoctor/asciidoctor#2921

But haven't managed to find a document that lists all of pandoc's AST supported elements.

**Comment by mb21** [comment]
@tajmone per definition (pun intended), see https://github.com/jgm/pandoc-types/blob/master/Text/Pandoc/Definition.hs#L95

**Comment by jgm** [comment]
Re-opening until we resolve some of the TODOs, but we now have a start on this in `doc/customizing-pandoc.md`.

**Comment by MattDodsonEnglish** [comment]
> But haven't managed to find a document that lists all of pandoc's AST supported elements.

When I was looking at the filter docs (related: https://github.com/jgm/pandoc/issues/8750), I start wondering how I could a reference of the AST. That led me to this issue.

I think I'm going to try to make it a hobby project to put together some AST docs. I can start by looking at the code in [`src/Text/Pandoc/Definition.hs`](https://github.com/jgm/pandoc-types/blob/master/src/Text/Pandoc/Definition.hs)

**Comment by jgm** [comment]
Isn't Text.Pandoc.Definition's haddock page a good canonical reference for the AST elements?
I'm curious what more you think would be needed in the way of documentation.

**Comment by tarleb** [comment]
One of the downsides of the [Haddock page](https://hackage.haskell.org/package/pandoc-types-1.23/docs/Text-Pandoc-Definition.html) is that is contains a lot of info and can be overwhelming. E.g., I'd guess that it's not immediately obvious to a user unfamiliar with Haskell that the list of instances can be skipped at first reading, but that the constructors are important.

**Comment by tajmone** [comment]
> I'm curious what more you think would be needed in the way of documentation.

I totally agree with @tarleb on the current docs being overwhelming to non-Haskell users.

## AST Spec as JSON/YAML Document within Pandoc Binary

Ideally, I'd love to see that pandoc would include a JSON or (YAML) file with the full AST specification (node name, type, attributes, etc.). If pandoc could auto-generate this JSON/YAML file (either within the source repository, or directly from the pandoc executable binary) and then provide a CLI command to emit it (e.g. `--print-ast-spec`) it would make life very easy, since it would be available without having to surf the web for that info.

The reason I think a JSON or YAML file would be better (i.e. rather than a markdown doc, etc.) is that while these formats are both is human-friendly enough to be consulted as they are, they can also be easily manipulated to create _ad hoc_ documents by parsing them and rendering them in whatever format one prefers (e.g. via Mustache templates). And, with the JSON/YAML spec being included within the executable, it would be very simple to setup any pandoc-related project to simply update the AST reference documentation with each new pandoc release by parsing it and re-generating the document via automated scripts.

**Comment by jgm** [comment]
> If pandoc could auto-generate this JSON/YAML file (either within the source repository, or directly from the pandoc executable binary) and then provide a CLI command to emit it (e.g. --print-ast-spec) it would make life very easy, since it would be available without having to surf the web for that info.

+1 on autogenerating.  What I want to avoid is a manually produced document that could get out of sync.

**Comment by MattDodsonEnglish** [comment]
Maybe for an "evergreen" strategy, it'd be possible to have:
- A generatated YAML reference
- A little conceptual explanation of what the tree is
- A rich set of examples of structure manipulation

------------

Note: I typed this before I saw the previous two responses in this thread, but @tarleb  and @tajmone  covered a lot of what I was thinking too. Here's the experience of a someone with only rudimentary programming knowledge and zero Haskell.

> Isn't Text.Pandoc.Definition's haddock page a good canonical reference for the AST elements?
I'm curious what more you think would be needed in the way of documentation.

Oh, interesting. I actually had scanned that but somehow didn't associate what I was looking at with a document structure. I might have understood if I'd opened the `Block` element, but the first element was `Pandoc`, which was too abstruse for me. Maybe I expected a tree/JSON representation, maybe with a little diagram of nodes (like how [MDN represents the DOM](https://developer.mozilla.org/en-US/docs/Web/API/Document_object_model/Using_the_Document_Object_Model)).

So, besides that it may start at too sharp a grade, I can't comment on that doc's usefulness since I haven't used it. :-) I'm going to experiment with using that as a reference to make some Lua filters and see how I do. 

I can speak of my experience as a Pandoc user at the lower end of technical proficiency, though. Basically, I'm only interested in transformation at the highest level: I just want a nice set of examples, a reference that I could look at, and maybe some example "lorem ipsum" style docs that I could inspect with `pandoc -t native`.

I don't know Haskell at all, so I can say that something like this 
`walk :: (Block -> Block) -> TableFoot -> TableFoot ` wasn't understandable to me. I suppose the reference examples assume Haskell knowledge: considering that Pandoc filtering is polyglottal, should that be necessary? Maybe a complementary, language-agnostic doc would make it easier for a wider audience to understand how to manipulate document structures.

 But, again, once I dig into those docs a bit, the reference will probably make more sense. I think Pandoc casts a pretty wide net (the Getting Started docs explains what `pwd` is), so I just posted this to document the experience of a reader with "fresh eyes."
